src/main/java/edu/kit/compiler/optimizations/PureFunctionOptimization.java

```diff
@@ -26,13 +26,15 @@ public boolean optimize(Graph graph, OptimizationState state) {
         for (var call : collector.calls) {
             var attributes = analysis.getAttributes(Util.getCallee(call));
             
-            hasChanged |= attributes.isPure() && unpinCall(call);
-
-            hasChanged |= switch (attributes.getPurity()) {
-                case CONST -> handleConstCall(call);
-                case PURE -> handlePureCall(call, collector.usedCalls); 
-                default -> false;
-            };
+            if (attributes.isTerminates()) {
+                hasChanged |= attributes.isPure() && unpinCall(call);
+
+                hasChanged |= switch (attributes.getPurity()) {
+                    case CONST -> handleConstCall(call);
+                    case PURE -> handlePureCall(call, collector.usedCalls); 
+                    default -> false;
+                };
+            }
```

Comment:
So we still flag non-terminating functions as pure/const, but we don't change the pin state / memory chain, right?

Reply:
Yes, that is the idea. The `purity` and `terminates` attributes are orthogonal to each other. Otherwise, we would introduce a bunch of illegal state into the attributes.